mlrun/runtimes/pod.py

```diff
@@ -411,6 +420,264 @@ def enrich_resources_with_default_pod_resources(
+                    NodeSelectorOperator.node_selector_op_not_in
+                ),
+            )
+            # remove affinity
+            self._prune_affinity_node_selector_requirement(
+                compile_affinity_by_label_selector(
+                    NodeSelectorOperator.node_selector_op_in
+                ),
+            )
+
+            # remove preemptible nodes constrain
+            self._prune_node_selector(mlconf.get_preemptible_node_selector())
+
+            # enrich with tolerations
+            self._enrich_with_tolerations(mlconf.get_preemptible_tolerations())
+
+            # TODO add gpu tolerations enrichment
+
+    def _initialize_affinity(self):
+        if not self.affinity:
+            self.affinity = client.V1Affinity()
+
+    def _initialize_node_affinity(self):
+        if not self.affinity.node_affinity:
+            self.affinity.node_affinity = client.V1NodeAffinity()
+
+    def _prune_affinity_node_selector_requirement(
+        self,
+        node_selector_requirements: typing.List[
+            kubernetes.client.V1NodeSelectorRequirement
+        ],
+    ):
+        """
+        Prunes given node selector requirements from affinity
+
+        :param node_selector_requirements:
+        :return:
+        """
+        # both needs to exist to prune required affinity from spec affinity
+        if not self.affinity or not node_selector_requirements:
+            return
+        if self.affinity.node_affinity:
+            node_affinity: client.V1NodeAffinity = self.affinity.node_affinity
+
+            new_required_during_scheduling_ignored_during_execution = None
+            # we are only looping over required_during_scheduling_ignored_during_execution
+            # because the scheduler can't schedule the Pod unless the rule is met
+            if node_affinity.required_during_scheduling_ignored_during_execution:
+                node_selector: client.V1NodeSelector = (
+                    node_affinity.required_during_scheduling_ignored_during_execution
+                )
+                new_node_selector_terms = self._get_node_selector_terms_without_provided_node_selector_requirements(
+                    node_selector_terms=node_selector.node_selector_terms,
+                    node_selector_requirements_to_remove=node_selector_requirements,
+                )
+                # check whether there are node selector terms to add to the new list of required terms
+                if len(new_node_selector_terms) > 0:
+                    new_required_during_scheduling_ignored_during_execution = (
+                        client.V1NodeSelector(
+                            node_selector_terms=new_node_selector_terms
+                        )
+                    )
+            # check if there is something to set into required_during_scheduling_ignored_during_execution
+            if (
+                not node_affinity.preferred_during_scheduling_ignored_during_execution
```

Comment:
How is the preferred related here ? not sure what you're doing here

Reply:
I intended to check if both are empty if true than clean node_affinity, I guess I missed it, fixed.